### DISCUSSION TITLE
Is there a way to initialize WGPU with HTMLCanvas/WebGlRenderingContext?

### DISCUSSION THREAD

**Discussion by Umenokin** [gfx-rs/wgpu#2893] [Q&A] [answered]
Hi folks!

As a way to learn rust, I want to migrate my TypeScript/WebGL2 pet project to WASM. I have WebGL2 context from HTML canvas and generally have no issues working directly with WebGL but it seems like working with WGPU will give me the ability to effortlessly switch to WebGPU once it goes live officially.

Is there a way to initialize WGPU in an HTML environment where I create and manage HTMLCanvas myself which means that I need to pass WebGlRenderingContext during WGPU initialization as well as manage events like canvas resizing myself?

Maybe there is a working example?

Thanks!

**Comment by grovesNL** [comment, marked as answer]
Hi @Umenokin! :wave:

You should be able to pass a `WebGlRenderingContext` using `instance.create_surface_from_canvas` https://github.com/gfx-rs/wgpu/blob/ea05b44f384077a1a466eeb9a58c523379225a01/wgpu/src/lib.rs#L1796 when you're using the wasm target.

**Reply by Umenokin** [reply]
Awesome, thanks @grovesNL !!! 😊

**Comment by ALPHAAAL** [comment]
Hi, I am trying to get wgpu working with HTML canvas as well.

Basically, what I am trying to achieve is
Start animation -> draw frame 1 - 1000 of the animation -> pause it -> do something on js side -> draw frames 2000 - 3000 (I intentionally skip frames, but it shouldn't matter).

I originally tried to use winit as the surface, however, I am not able to figure out how to pause the animation and start it again without creating new event loop (winit doesn't allow creating multiple event loops).

I tried to use `create_surface_from_canvas` as mentioned by @grovesNL in this thread, however, it seems like the frame rate is much lower than using winit for some reason, and I am not able to figure out why.

Any guidance would be nice, thanks! 😭

**Comment by anthonydito** [comment]
As of Mar. 2024, the accepted answer isn't up to date with the newest version of wgpu. To create a surface from a canvas, you can do.

```rust
let html_canvas: HtmlCanvasElement = ...
let instance = wgpu::Instance::new(InstanceDescriptor {
    backends: wgpu::Backends::all(),
    ..InstanceDescriptor::default()
});
let surface_target = SurfaceTarget::Canvas(html_canvas);
let surface = instance.create_surface(surface_target)
```

**Reply by samuelselleck** [reply]
For anyone else getting stuck on this: `SurfaceTarget::Canvas` is under the `#[cfg(target_arch="wasm32")]` flag. If you haven't configured your LSP to use this `SurfaceTarget::Canvas` will look like it's not defined, but it runs fine when compiling.

**Reply by SonyStone** [reply]
> configured your LSP

Hello, I am the one who get stuck on this `SurfaceTarget::Canvas`.
Maybe you know how to configure rust analyzer in vscode for web build?

Nevermind, I found in vscode setting where I added target and now it's working
```json
{
    "rust-analyzer.cargo.target": "wasm32-unknown-unknown"
}
```

**Reply by KySpace** [reply]
Does it configure for all projects or the current project? I don't want to change it for everything, though.

**Reply by cwfitzgerald** [reply]
When editing the setting you can choose to set it either for your user or for the workspace. The workspace takes precedent over any user settings.

**Reply by KySpace** [reply]
I'm now doing it in a leptos crate. Is there a way I can enable `wasm32` everywhere in the project?

**Reply by KySpace** [reply]
So I followed this method, and I think the surface is created, but I got `Surface is not configured for presentation` when trying to generate the surface texture.

I'm mostly following the [tutorial](https://sotrh.github.io/learn-wgpu/beginner/tutorial2-surface/#first-some-housekeeping-state). Maybe there's something else that I should change?

**Reply by KySpace** [reply]
Oh, I got it. I need to call `surface.configure(&device, &config)` first.
